### ISSUE TITLE
Nested anchor elements can end up nested in generated DOM tree

### ISSUE THREAD

**Issue by cketti** [jhy/jsoup#845]
This is a condensed version of the issue reported in https://github.com/jhy/jsoup/issues/728#issuecomment-286945987

**Input:**

```html
<html>
  <body>
    <a href='#1'>
      <div>
        <div>
          <a href='#2'>child</a>
        </div>
      </div>
    </a>
  </body>
</html>
```

**Output:** Jsoup (f28c024ba127fd701f0d195a359afbabff04d7a1)

```html
<html>
 <head></head>
 <body> 
  <a href="#1"> </a>
  <div>
   <a href="#1"> 
    <div> 
     <a href="#2">child</a> 
    </div></a> 
  </div>   
 </body>
</html>
```

**Expected:** Chrome transforms the input into this:
```html
<html><head></head><body>
    <a href="#1">
      </a><div><a href="#1">
        </a><div><a href="#1">
          </a><a href="#2">child</a>
        </div>
      </div>
</body></html>
```

The DOM tree generated by Jsoup contains nested anchor elements which are not allowed.
Having only one nested `div` element in the input seems to do the right thing. At least two nested block elements are necessary to trigger this bug.

**Comment by jhy** [comment]
I believe this is stemming from jsoup's implementation of the [adoption agency algorithm](https://html.spec.whatwg.org/multipage/parsing.html#adoption-agency-algorithm), which is in [HtmlTreeBuilderState](https://github.com/jhy/jsoup/blob/master/src/main/java/org/jsoup/parser/HtmlTreeBuilderState.java#L555)).

(The impl is in end-tag handling while the spec shows it on the start tag, but that's because an implicit end-tag event is generated.)

I reviewed it yesterday a bit but start going cross-eyed. The spec is worded in a way that I was having troubles following when skimming looking for a bug, and it looks like it was updated a bit and moved around since when I first implemented it.

My guess is that the essence of the bug is that we're working from the bottom of the list and stopping too early. But could be pretty off track here. I really need to review the whole flow. Or hopefully this gives enough of a starting point to somebody else, in which case I'll offer 25 imaginary internet points if you fix it ;)
